substrate/primitives/arithmetic/src/fixed_point.rs

```diff
@@ -983,9 +983,98 @@ macro_rules! implement_fixed {
 			type Err = &'static str;
 
 			fn from_str(s: &str) -> Result<Self, Self::Err> {
-				let inner: <Self as FixedPointNumber>::Inner =
-					s.parse().map_err(|_| "invalid string input for fixed point number")?;
-				Ok(Self::from_inner(inner))
+				let s = s.trim();
+
+				// Check if the string contains a decimal point
+				if let Some(dot_pos) = s.find('.') {
+					// Parse as decimal number (e.g., "1.0", "123.456")
+					let (integer_part, fractional_part) = s.split_at(dot_pos);
+					let fractional_part = &fractional_part[1..]; // Skip the '.'
+
+					// Check if it's a negative number
+					let is_negative = integer_part.starts_with('-');
+
+					// For unsigned types, reject negative numbers
+					if is_negative && !$name::SIGNED {
+						return Err("negative numbers not supported for unsigned fixed point types");
+					}
+
+					// Parse integer part
+					let integer: i128 = if integer_part.is_empty() {
+						0
+					} else {
+						integer_part
+							.parse()
+							.map_err(|_| "invalid integer part in decimal number")?
+					};
+
+					// Parse fractional part
+					let fractional_raw: u128 = if fractional_part.is_empty() {
+						0
+					} else {
+						fractional_part
+							.parse()
+							.map_err(|_| "invalid fractional part in decimal number")?
+					};
+
+					// Convert fractional part to the appropriate scale
+					let fractional_digits = fractional_part.len() as u32;
+
+					// Calculate fractional value using more careful arithmetic
+					let fractional_scaled: i128 = if fractional_digits > 0 {
+						// Use saturating arithmetic to avoid overflow, then check bounds
+						let scale_factor = 10u128.saturating_pow(fractional_digits);
+						if scale_factor == 0 {
+							return Err("fractional part has too many digits");
+						}
```

Comment:
```suggestion
						// Checked math to avoid overflow
						let Some(scale_factor) = 10u128.checked_pow(fractional_digits) else {
							return Err("fractional part has too many digits");
						};
```
And why check for `0`? Should it not saturate at `i128::MAX`?